HDDS-13112. [Docs] OM Bootstrap can also happen when follower falls behind too much. (#8600)
apache/ozone: hadoop-hdds/docs/content/feature/OM-HA.md

Note that using the _force_ option during bootstrap could crash the OM process if it does not have updated configurations.

## Automatic Snapshot Installation for Stale Ozone Managers

> Sometimes an OM follower node may be offline or fall so far behind the leader OM's log that it cannot catch up by replaying individual log entries.  The OM HA implementation includes an automatic snapshot installation and recovery process for such cases.

Reviewer: Since git diff compares files line by line, I suggest using short lines for better tracking.  I also revise some sentenses.

```md
Sometimes an OM follower node may be offline or fall far behind the OM leader's raft log.
Then, it cannot easily catch up by replaying individual log entries.
The OM HA implementation includes an automatic snapshot installation 
and recovery process for such cases.
```

Reviewer 2: Try `git diff --word-diff`, it works well for text and can clarify changes in long lines.

Unless the change is trivial (e.g. fixing a typo), short lines usually either get unbalanced, or rewrapping hides the actual change.  Then `git diff` with or without `--word-diff` becomes less useful.

Example word diff from recent doc change (as a screenshot, because GitHub does not support highlighting word diff properly, but Git does):

![image](https://github.com/user-attachments/assets/b464cc43-cab2-446c-9685-11b77adac1a6)

How it works:

1. Leader determines that the follower is too far behind.

> 2. Leader notifies the follower to catch up via snapshot.

Reviewer: Let's change it to `... to install a snapshot.`

> 3. The follower downloads and installs the latest snapshot from the leader.

Reviewer: > ...  the latest snapshot from the leader.

Side note: we should support install snapshot from another follower in order to reduce the load of the leader.

Author: https://issues.apache.org/jira/browse/HDDS-13267

4. After installing the snapshot, the follower OM resumes normal operation and log replication from the new state.

> This logic is implemented in the [`OzoneManagerStateMachine.notifyInstallSnapshotFromLeader()` method](https://github.com/apache/ozone/blob/931bc2d8a9e8e8595bb49034c03c14e2b15be865/hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/ratis/OzoneManagerStateMachine.java#L521-L541).

Reviewer: Let's point to a particular release:
```md
This logic is implemented in the `OzoneManagerStateMachine.notifyInstallSnapshotFromLeader()`;
see the [code](https://github.com/apache/ozone/blob/ozone-2.0.0/hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/ratis/OzoneManagerStateMachine.java#L520-L531)
in Release 2.0.0.
```

It would be great if there is a way to specify a base URL variable something like below:
```
BASE_URL=https://github.com/apache/ozone/blob/ozone-2.0.0
```
Then the link becomes
```md
[code](${BASE_URL}/hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/ratis/OzoneManagerStateMachine.java#L520-L531)
```

> In most scenarios, stale OMs will recover automatically, even if they have missed a large number of operations.  Manual intervention (such as running `ozone om --bootstrap`) is only required when adding a new OM node to the cluster or when explicitly requested by support instructions.

Reviewer 3: AFAIK `ozone om --bootstrap` is only used for adding an OM and cannot be used as a "Manual intervention" to catch up with the leader.

The way we did it in our side is to use clean the Ratis logs directory and OM DB and resync the OM DB (using Rsync). 

1. stop om follower
ozone --daemon stop om
 
2. backup metadata
timestamp=$(date '+%Y-%m-%dT%H_%M')
mv /hadoop/ozone/meta/om/db/om.db /hadoop/ozone/meta/om/db/om.db.bak_${timestamp}
mv /hadoop/ozone/meta/om/ratis /hadoop/ozone/meta/om/ratis_${timestamp}
mv /hadoop/ozone/meta/om/ratis-snapshot /hadoop/ozone/meta/om/ratis-snapshot_${timestamp}
mkdir /hadoop/ozone/meta/om/ratis  /hadoop/ozone/meta/om/ratis-snapshot

3. Change the ozone.om.ratis.log.purge.preservation.log.num=0 to prevent the bug described in RATIS-2177 after startup (note the original value since the configuration will need to be reverted)

4. rsync OM leader's meta to follower  （Execute multiple times to ensure data consistency）
rsync -av --delete /hadoop/ozone/meta/om/db/om.db ${OM_FOLLOWER_ADDRESS}:/hadoop/ozone/meta/om/db/

4. restart om follower
ozone --daemon start om

5. Wait until the first "purge" in OM log
   2024-11-08 08:01:54,317 [om4@group-13A745F1EB59-StateMachineUpdater] INFO org.apache.ratis.server.raftlog.RaftLog: om4@group-13A745F1EB59-SegmentedRaftLog: purge 88499622476

6. Change the ozone.om.ratis.log.purge.preservation.log.num back to the original value

7. restart om follower again (ensure that all 3 OMs are up and the current om is not a leader)
   ozone --daemon stop om; sleep 3; ozone --daemon start om

Reviewer 3: This doesn't include OM snapshot and snapdiff DB, which might need to be handled by tarring the whole OM DB instead of using Rsync (to preserve with hard links)

Author: We use snapshot checkpointing for metadata backup purposes. But let me remove the sentence "when explicitly requested by support instructions." because it does reads ambiguous.

> 

Reviewer 4: ```suggestion
Important Note on Ozone Manager (OM) Disk Space for Snapshots
When an Ozone Manager (OM) acts as a follower in an HA setup, it downloads snapshot tarballs from the leader to its local metadata directory. Therefore, always ensure your OM disks have at least 2x the current OM database size to accommodate the existing data and incoming snapshots, preventing disk space issues and maintaining cluster stability.
```
We could add this as a note.

## References

 * Check [this page]({{< ref "design/omha.md" >}}) for the links to the original design docs
 * Ozone distribution contains an example OM HA configuration, under the `compose/ozone-om-ha` directory which can be tested with the help of [docker-compose]({{< ref "start/RunningViaDocker.md" >}}).

> * [Apache Ratis State Machine API documentation](https://github.com/apache/ratis/blob/3612bcaf7d3e48a658935fc8b250e5d3b35df174/ratis-server-api/src/main/java/org/apache/ratis/statemachine/StateMachine.java)

Reviewer: Use a release tag:
https://github.com/apache/ratis/blob/ratis-3.1.3/ratis-server-api/src/main/java/org/apache/ratis/statemachine/StateMachine.java

Note that using the _force_ option during bootstrap could crash the OM process if it does not have updated configurations.

## Automatic Snapshot Installation for Stale Ozone Managers

> In an Ozone Manager (OM) High Availability (HA) cluster, all OM nodes maintain a consistent metadata state using the Ratis consensus protocol. Sometimes, an OM follower node may be offline or fall so far behind the leader OM’s log that it cannot catch up by replaying individual log entries.
> 
> The OM HA implementation includes an **automatic snapshot installation and recovery process** for such cases:
> 
> - **Snapshot Installation Trigger:**  
> When a follower OM falls significantly behind and is unable to catch up with the leader OM through standard log replication, the Ratis consensus layer on the leader OM may determine that a snapshot installation is necessary. The leader then notifies the follower, and the snapshot installation on the follower is handled by its `OzoneManagerStateMachine`.
> 
> - **How it works:**
>   - The follower OM receives a snapshot installation notification from the leader via the consensus protocol.
>   - The follower OM then downloads and installs the latest consistent checkpoint (snapshot) from the leader OM.
>   - After installing the snapshot, the follower OM resumes normal operation and log replication from the new state.
> 
> - **Relevant Implementation:**  
>   This logic is implemented in the `OzoneManagerStateMachine.notifyInstallSnapshotFromLeader()` method. The install is triggered automatically by the consensus layer (Ratis) when it detects that a follower cannot catch up by log replay alone.
> 
> - **What this means for administrators:**
>   - In most scenarios, stale OMs—whether they were temporarily offline or simply fell too far behind the leader while remaining online—will recover automatically, even if they have missed a large number of operations.
>   - Manual intervention (such as running `ozone om --bootstrap`) is only required when adding a new OM node to the cluster or when explicitly requested by support instructions.
> 

Reviewer 2: - don't repeat yourself
- "cases" are not listed after this, so ":" at the end is not appropriate
- use permalink for code, otherwise it will point to the wrong location after changes to `OzoneManagerStateMachine`

```suggestion
Sometimes an OM follower node may be offline or fall so far behind the leader OM's log that it cannot catch up by replaying individual log entries.  The OM HA implementation includes an automatic snapshot installation and recovery process for such cases.

How it works:

1. Leader determines that the follower is too far behind.
2. Leader notifies the follower to catch up via snapshot.
3. The follower downloads and installs the latest snapshot from the leader.
4. After installing the snapshot, the follower OM resumes normal operation and log replication from the new state.

This logic is implemented in the [`OzoneManagerStateMachine.notifyInstallSnapshotFromLeader()` method](https://github.com/apache/ozone/blob/931bc2d8a9e8e8595bb49034c03c14e2b15be865/hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/ratis/OzoneManagerStateMachine.java#L521-L541).

In most scenarios, stale OMs will recover automatically, even if they have missed a large number of operations.  Manual intervention (such as running `ozone om --bootstrap`) is only required when adding a new OM node to the cluster or when explicitly requested by support instructions.
```

## References

 * Check [this page]({{< ref "design/omha.md" >}}) for the links to the original design docs
 * Ozone distribution contains an example OM HA configuration, under the `compose/ozone-om-ha` directory which can be tested with the help of [docker-compose]({{< ref "start/RunningViaDocker.md" >}}).

> * [OzoneManagerStateMachine.notifyInstallSnapshotFromLeader source code](https://github.com/apache/ozone/blob/master/hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/ratis/OzoneManagerStateMachine.java#L530)

Reviewer 2: Suggesting inline link.

```suggestion
```

> * [Apache Ratis State Machine API documentation](https://github.com/apache/ratis/blob/master/ratis-server-api/src/main/java/org/apache/ratis/statemachine/StateMachine.java)

Reviewer 2: ```suggestion
* [Apache Ratis State Machine API documentation](https://github.com/apache/ratis/blob/3612bcaf7d3e48a658935fc8b250e5d3b35df174/ratis-server-api/src/main/java/org/apache/ratis/statemachine/StateMachine.java)
```
